### PR TITLE
Can clean right after playback stops

### PR DESCRIPTION
**Pull request by kugan49**
This one is ok for you ?

Note that I've not activated clean on library update ( mark a video as seen)

just uncomment line 82 and comment line 83 to activate this

### CODE REVIEW

**Comment by Anthirian** [comment]
Thanks, this way of sending your pull request is fine. I'll review it soon, but since it's quite a large one it might take me a while.

**Comment by Anthirian** [comment]
I have finally found the time to look at your pull request. To me it looks as if it is not yet fully finished. By that I mean you have added some code for the functionality, but I get the feeling you have fully tested it to see if it is bug free. My IDE shows me several code warnings. Furthermore I have a personal preference for named arguments with default values instead of using ``**kwargs`` so I know exactly which arguments are (or can be) used to call the ``clean()`` method for example. Also I think it would be much clearer if there would be a standalone method for cleaning specific files this way, for example ``clean_file(video_type, file)`` instead of adding more optional functionality to the existing ``clean_all()`` and ``clean()`` methods.

I do like the fact that the option depends on the fact whether the cleaning service is enabled or not. This means that the user can, in addition to interval-based cleaning, opt for even quicker cleaning. Though maybe if the user enables instant cleaning, the interval timer should be reset as well every time a file is cleaned.

As of now I don't think it's ready to be merged just yet. I appreciate the effort you put into this, but I'd like to keep the code very consistent and clean. Please take a look at the code comments and update your commit accordingly by doing a forced push and I will review it again.

File: default.py [hunk 1]
```diff
@@ -102,7 +101,7 @@ def hide_progress(self):
         """
         self.silent = True
 
-    def clean(self, video_type):
+    def clean(self, video_type, **kwargs ):
```
Comment (Anthirian) [inline default.py]: I would like to have the header be ``def clean(self, video_type, expired_videos=None)`` as explained in the main comment thread. This saves having to add the ``kwargs.get('expired_videos', None)`` line and makes the allowed arguments look more clear to callers of the function. However, as I also explained, I would like to have a separate function for cleaning single videos as triggered by the monitor.

File: default.py [hunk 2]
```diff
@@ -111,6 +110,11 @@ def clean(self, video_type):
         :rtype: (list, int, int)
         :return: A list of the filenames that were cleaned, as well as the number of files cleaned and the return status.
         """
+        expired_videos=kwargs.get('expired_videos', None)
+        if expired_videos!=None:
```
Comment (Anthirian) [inline default.py]: I prefer ``if expired_videos:``. You can even rewrite these lines to ``immediate_file_clean = True if expired_videos else False`` for short.

File: default.py [hunk 3]
```diff
@@ -111,6 +110,11 @@ def clean(self, video_type):
...
+            ImmediateFileClean = True
+        else:
+            ImmediateFileClean = False
```
Comment (Anthirian) [inline default.py]: Please do not use CamelCase for variable names. I use ``lowercase_with_underscores`` throughout the code.

File: default.py [hunk 4]
```diff
@@ -204,32 +209,41 @@ def clean(self, video_type):
 
         return cleaned_files, count, self.exit_status
 
-    def clean_all(self):
+    def clean_all(self, **kwargs):
```
Comment (Anthirian) [inline default.py]: See line 104.

File: default.py [hunk 5]
```diff
@@ -204,32 +209,41 @@ def clean(self, video_type):
...
+        
+        
         if get_setting(clean_when_idle) and xbmc.Player().isPlaying():
             debug("Kodi is currently playing a file. Skipping cleaning.", xbmc.LOGWARNING)
             return None, self.exit_status
 
         results = {}
         cleaning_results, cleaned_files = [], []
-        if not get_setting(clean_when_low_disk_space) or (get_setting(clean_when_low_disk_space) and disk_space_low()):
+        if (expired_videos != none) or (not get_setting(clean_when_low_disk_space) or (get_setting(clean_when_low_disk_space) and disk_space_low())):
```
Comment (Anthirian) [inline default.py]: This is one of the lines that triggered a warning in my IDE. You did not capitalize ``None`` and as such you are comparing it to the (nonexistent) variable ``none``, which will most likely result in a runtime error. Also, comparing with ``None`` is not necessary. Any value other than the empty dict ``{}`` will evaluate to ``True`` so you can just write ``if expired_videos:``

File: default.py [hunk 6]
```diff
@@ -204,32 +209,41 @@ def clean(self, video_type):
...
                 self.progress.create(ADDON_NAME, *map(translate, (32619, 32615, 32615)))
                 self.progress.update(0)
                 self.monitor.waitForAbort(2)
-            for video_type in [self.MOVIES, self.MUSIC_VIDEOS, self.TVSHOWS]:
-                if not self.__is_canceled():
-                    cleaned_files, count, status = self.clean(video_type)
-                    if count > 0:
-                        cleaning_results.extend(cleaned_files)
-                        results[video_type] = count
+            if expired_videos != none :
```
Comment (Anthirian) [inline default.py]: See the comments for lines 104 and 230.

File: service.py [hunk 7]
```diff
@@ -37,9 +38,69 @@ def autostart():
         else:
             cleaner.monitor.waitForAbort(service_sleep)
 
-    log("Abort requested. Terminating.")
+    debug("Abort requested. Terminating.")
```
Comment (Anthirian) [inline service.py]: Well spotted. I apparently forgot to change all ``log()``s back to ``debug()``s.

File: service.py [hunk 8]
```diff
@@ -37,9 +38,69 @@ def autostart():
...
     return
 
 
+
```
Comment (Anthirian) [inline service.py]: I'm gonna write a comment here about the entire ``service.py`` file, so it's easier to read. I like the use of a ``Monitor`` to listen for notifications from the player. I have added some general comments regarding improvement below.

- I would like to ask that you use double quotes instead of single quotes, to make the code more consistent. 
- The same I would ask about using format strings: please use ``"This kind of {0} notation".format("string formatting")`` instead of ``"this %s of formatting" % "type"``. 
- Finally I would like to ask that you write detailed and specific debugging messages (in English). This helps me immensely in case a future update breaks something in your code and you are not around to help.

File: service.py [hunk 9]
```diff
@@ -37,9 +38,69 @@ def autostart():
...
+        except:
+            debug('failed to do janitor on video %s (filename= %s )' % (expired_videos['filename'],expired_videos['title']), xbmc.LOGNOTICE)
+        
+   
+# monitor notifications
+class MyPlayer(xbmc.Monitor):
+    def __init__( self, *args, **kwargs ):
+        debug('Player Class Init')
+        xbmc.Monitor.__init__( self )
+        self.action = kwargs['action']
```
Comment (Anthirian) [inline service.py]: Here I would again use ``def __init__(self, action)`` and if required with a default value.

File: service.py [hunk 10]
```diff
@@ -37,9 +38,69 @@ def autostart():
         else:
             cleaner.monitor.waitForAbort(service_sleep)
 
-    log("Abort requested. Terminating.")
+    debug("Abort requested. Terminating.")
     return
 
 
+    
+class Main:
```
Comment (Anthirian) [inline service.py]: This an old-style class definition. You should use ``class Main()`` instead.

Quick question: is it really necessary to create this class just to create the ``MyPlayer`` object? I have not used the ``xbmc.Monitor`` class before.

File: service.py [hunk 11]
```diff
@@ -37,9 +38,69 @@ def autostart():
...
+    def __init__( self ):
+        debug('init main')
+        self._service_setup()
+        while (not xbmc.abortRequested):
```
Comment (Anthirian) [inline service.py]: Braces are not required here.

File: service.py [hunk 12]
```diff
@@ -37,9 +38,69 @@ def autostart():
...
+        debug('installing service')
+        self.Player = MyPlayer(action = self._service_janitor)
+
+    def _service_janitor( self, video_type, expired_videos ):
+        debug('service janitor')
+        # mark as watched
+        debug('%s to clean : %s (%s)' % (video_type, expired_videos['title'], expired_videos['filename']),xbmc.LOGNOTICE)
+        try:
+            cleaner = Cleaner()
+            results, _ = cleaner.clean_all(video_type=video_type, expired_videos=expired_videos)
```
Comment (Anthirian) [inline service.py]: I would like to see this method call the (to be created) ``clean_file(video_type, file)`` instead of having function creep on ``clean_all()`` and ``clean()``, as I explained in the main comment thread. Also, is there a particular reason you're not interested in the return status? It would help you determine if the cleaning was successful or not and go from there.

File: service.py [hunk 13]
```diff
@@ -37,9 +38,69 @@ def autostart():
...
+            # do janitor
+            debug(results)
+            notify(results)
```
Comment (Anthirian) [inline service.py]: If you do a call to ``notify()``, this will call ``debug()`` automatically, so that all notification are also always logged. Doing both is redundant here.

File: service.py [hunk 14]
```diff
@@ -37,9 +38,69 @@ def autostart():
...
+        except:
```
Comment (Anthirian) [inline service.py]: This will catch all exceptions that occur, even ones that have nothing to do with the instant cleaning. Please specify exactly which exception you wish to catch and act upon. Other exceptions can then be caught elsewhere.

File: service.py [hunk 15]
```diff
@@ -37,9 +38,69 @@ def autostart():
...
+                
+                debug('xbmc %s' % method)
+                result = json.loads(data)
+                if method == 'Player.OnStop' or ( method == 'VideoLibrary.OnUpdate' and 'playcount' in result ):
+                    # if viewing in file mode and playback stopped at the end
+                    if 'item' in result and 'title' in result["item"] and result["end"]:
+                        if result['item']['type'] == 'episode' or result['item']['type'] == 'movie' or result['item']['type'] == 'musicvideo':
+                            # mark episode as watched
+                            
+                            xpired_videos = {'filename':result["item"]["file"],'title':result["item"]["title"]}
```
Comment (Anthirian) [inline service.py]: You mistyped the variable name, causing a warning in my IDE.

File: service.py [hunk 16]
```diff
@@ -37,9 +38,69 @@ def autostart():
...
+                            self.action(result['item']['type'] + 's',expired_videos)
+                    elif method == 'Player.OnStop':
+                        # wait 1s before marking episode
+                        xbmc.sleep(1000)
+
 if __name__ == "__main__":
-    autostart()
+    debug('chargement main service')
```
Comment (Anthirian) [inline service.py]: As explained, please only use English for ``debug()`` and localize all text that uses ``notify()``.

File: settings.py [hunk 17]
```diff
@@ -76,7 +77,11 @@ def get_setting(setting):
     elif setting in strings:
         return str(utils.ADDON.getSetting(setting))
     elif setting in paths:
-        return translatePath(utils.ADDON.getSetting(setting).encode("utf-8"))
+        try:
```
Comment (Anthirian) [inline settings.py]: This is part of your previous pull request. I have already changed the code for this part, so please revert to that solution and include it in your updated commit so that this no longer is part of the pull request.

File: addon.xml [hunk 18]
```diff
@@ -3,6 +3,7 @@
     <requires>
         <import addon="xbmc.python" version="2.25.0" />
         <import addon="xbmc.json" version="7.0.0" />
+        <import addon="script.module.simplejson" version="2.0.10" />
```
Comment (Anthirian) [inline addon.xml]: There is no need for importing simplejson anymore. I make use of ``import json`` in ``default.py``, which is actually a speed improvement for low-resource systems such as the Raspberry Pi.

File: service.py (hunk 10 above)
Comment (kugan49) [inline service.py]: I Don't know I take this code from an other script I use and I know it works ;)

**Comment by kugan49** [comment]
This is done.
Tell me if there is other problems.

**Comment by Anthirian** [comment]
I finally had some more time to check out the changes you made. I still get the feeling you haven't fully tested the code yourself, because now I am getting lots of different errors.

I'm sort of okay with the `Monitor` class, but it needs some small tweaks to make it bug free it seems. I can take care of that myself. I'm not too sure about the `work_on_file` method though. I notice the indenting is not correct and I have lots of unresolved references if I fix the indenting. It sort of looks like you shifted some code around and gave it a new name.

In short, I think I'm going to use your Monitor and modify the cleaning functions to support single file cleaning when I have the time. I'll leave this open until then, but feel free to update the code after you've fully tested it and made the necessary changes. I do still value you taking the time to help out, but I really don't want to introduce new bugs that I need to fix. 😄

**Comment by kugan49** [comment]
Update from 4.4.0 and review you sayed. I you are always interested.

**Comment by kugan49** [comment]
Finally can't make work as you want so I close
